// @sanity/ui/dist/cjs/components/autocomplete/styles.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  ListBoxCard: () => ListBoxCard,
  ListBoxContainer: () => ListBoxContainer,
  LoadingCard: () => LoadingCard,
  Root: () => Root
});
var import_styled_components = __toModule(require("styled-components"));
var import_primitives = __toModule(require("../../primitives"));
const Root = import_styled_components.default.div`
  position: relative;
`;
const ListBoxContainer = import_styled_components.default.div`
  position: relative;
`;
const ListBoxCard = (0, import_styled_components.default)(import_primitives.Card)`
  position: absolute;
  top: 0;
  left: 1px;
  right: 1px;
  z-index: 1000;
  max-height: calc(100vh - 10em);
  overflow: auto;
  -webkit-overflow-scrolling: touch;
  min-height: 35px;

  & > ul {
    list-style: none;
    padding: 0;
    margin: 0;
  }
`;
const LoadingCard = (0, import_styled_components.default)(import_primitives.Card)`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  pointer-events: none;
  align-items: center;
  justify-content: center;
  transition: opacity 100ms;

  &:not([hidden]) {
    display: flex;
  }
`;
