// @sanity/ui/dist/cjs/primitives/grid/grid.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  Grid: () => Grid
});
var import_react = __toModule(require("react"));
var import_styled_components = __toModule(require("styled-components"));
var import_internal = __toModule(require("../../styles/internal"));
var import_box = __toModule(require("../box"));
const Root = (0, import_styled_components.default)(import_box.Box)(import_internal.responsiveGridStyle);
const Grid = (0, import_react.forwardRef)((props, ref) => {
  const {
    as,
    autoRows,
    autoCols,
    autoFlow,
    columns,
    gap,
    gapX,
    gapY,
    rows,
    children
  } = props, restProps = __rest(props, [
    "as",
    "autoRows",
    "autoCols",
    "autoFlow",
    "columns",
    "gap",
    "gapX",
    "gapY",
    "rows",
    "children"
  ]);
  return /* @__PURE__ */ import_react.default.createElement(Root, __assign(__assign({
    "data-as": typeof as === "string" ? as : void 0,
    "data-ui": "Grid"
  }, restProps), {
    $autoRows: autoRows,
    $autoCols: autoCols,
    $autoFlow: autoFlow,
    $columns: columns,
    $gap: gap,
    $gapX: gapX,
    $gapY: gapY,
    $rows: rows,
    forwardedAs: as,
    ref
  }), children);
});
Grid.displayName = "Grid";
