// @sanity/form-builder/lib/inputs/files/common/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Overlay = exports.FileTarget = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _ui = require("@sanity/ui");

var _fileTarget = require("../../common/fileTarget");

var _focusringUtils = require("./focusringUtils");

var _templateObject, _templateObject2;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var FileTarget = (0, _styledComponents.default)((0, _fileTarget.fileTarget)(_ui.Card))((_ref) => {
  var theme = _ref.theme;
  var border = {
    width: 1,
    color: 'var(--card-border-color)'
  };
  return (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    --card-focus-box-shadow: ", ";\n\n    border-radius: ", ";\n    outline: none;\n    box-shadow: var(--card-focus-box-shadow);\n\n    &:focus {\n      --card-focus-box-shadow: ", ";\n    }\n  "])), (0, _focusringUtils.focusRingBorderStyle)(border), (0, _ui.rem)(theme.sanity.radius[1]), (0, _focusringUtils.focusRingStyle)({
    base: theme.sanity.color.base,
    border,
    focusRing: theme.sanity.focusRing
  }));
});
exports.FileTarget = FileTarget;

var Overlay = _styledComponents.default.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  background-color: var(--card-bg-color);\n  z-index: 3;\n  pointer-events: none;\n"])));

exports.Overlay = Overlay;
