// @sanity/dashboard/lib/DashboardTool.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Dashboard = _interopRequireDefault(require("./containers/Dashboard"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var strokeStyle = {
  stroke: 'currentColor',
  strokeWidth: 1.2
};

var DashboardIcon = () => /*#__PURE__*/_react.default.createElement("svg", {
  "data-sanity-icon": true,
  viewBox: "0 0 25 25",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg",
  preserveAspectRatio: "xMidYMid",
  width: "1em",
  height: "1em"
}, /*#__PURE__*/_react.default.createElement("path", {
  d: "M19.5 19.5H5.5V5.5H19.5V19.5Z",
  style: strokeStyle
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M5.5 12.5H19.5",
  style: strokeStyle
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M14.5 19.5V12.5M10.5 12.5V5.5",
  style: strokeStyle
}));

var _default = {
  title: 'Dashboard',
  name: 'dashboard',
  icon: DashboardIcon,
  component: _Dashboard.default
};
exports.default = _default;
