// @sanity/base/lib/theme/color.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.color = void 0;

var _color = require("@sanity/color");

var _ui = require("@sanity/ui");

var _helpers = require("./helpers");

var _legacyPalette = require("./legacyPalette");

var _legacyTones = require("./legacyTones");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var color = (0, _ui.createColorTheme)({
  base: (_ref) => {
    var navbar = _ref.dark,
        name = _ref.name;
    var stateTones = navbar ? _legacyTones.legacyTones.state.navbar : _legacyTones.legacyTones.state.default;
    var dark = stateTones.dark;

    if (name === 'default') {
      return {
        fg: stateTones.fg,
        bg: stateTones.bg,
        border: stateTones.default[200],
        focusRing: _legacyPalette.legacyPalette.focus.base,
        shadow: {
          outline: (0, _ui.rgba)(stateTones.default[500], dark ? 0.2 : 0.4),
          umbra: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : stateTones.default[500], 0.2),
          penumbra: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : stateTones.default[500], 0.14),
          ambient: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : stateTones.default[500], 0.12)
        }
      };
    }

    if (name === 'transparent') {
      var _tints = stateTones.default;
      return {
        fg: _tints[900],
        bg: _tints[50],
        border: _tints[300],
        focusRing: _legacyPalette.legacyPalette.focus.base,
        shadow: {
          outline: (0, _ui.rgba)(_tints[500], dark ? 0.2 : 0.4),
          umbra: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : _tints[500], 0.2),
          penumbra: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : _tints[500], 0.14),
          ambient: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : _tints[500], 0.12)
        }
      };
    }

    var tints = stateTones[name] || stateTones.default;
    return {
      fg: tints[900],
      bg: tints[50],
      border: tints[200],
      focusRing: tints[500],
      shadow: {
        outline: (0, _ui.rgba)(tints[500], dark ? 0.2 : 0.4),
        umbra: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : tints[500], 0.2),
        penumbra: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : tints[500], 0.14),
        ambient: (0, _ui.rgba)(dark ? _legacyPalette.legacyPalette.black : tints[500], 0.12)
      }
    };
  },
  solid: (_ref2) => {
    var base = _ref2.base,
        navbar = _ref2.dark,
        state = _ref2.state,
        tone = _ref2.tone;
    var buttonTones = navbar ? _legacyTones.legacyTones.button.navbar : _legacyTones.legacyTones.button.default;
    var dark = buttonTones.dark;
    var blend = dark ? _helpers._screen : _helpers._multiply;
    var tints = buttonTones[tone] || buttonTones.default;

    if (state === 'disabled') {
      return {
        bg: blend(base.bg, buttonTones.default[200]),
        border: blend(base.bg, buttonTones.default[200]),
        fg: blend(base.bg, buttonTones.bg)
      };
    }

    if (state === 'hovered') {
      return {
        bg: blend(base.bg, tints[600]),
        border: blend(base.bg, tints[600]),
        fg: blend(base.bg, buttonTones.bg)
      };
    }

    if (state === 'pressed') {
      return {
        bg: blend(base.bg, tints[800]),
        border: blend(base.bg, tints[800]),
        fg: blend(base.bg, buttonTones.bg)
      };
    }

    if (state === 'selected') {
      return {
        bg: blend(base.bg, tints[800]),
        border: blend(base.bg, tints[800]),
        fg: blend(base.bg, buttonTones.bg)
      };
    } // state: "enabled" | unknown


    return {
      bg: blend(base.bg, tints[500]),
      border: blend(base.bg, tints[500]),
      fg: blend(base.bg, buttonTones.bg)
    };
  },
  muted: (_ref3) => {
    var base = _ref3.base,
        navbar = _ref3.dark,
        state = _ref3.state,
        tone = _ref3.tone;
    var stateTones = navbar ? _legacyTones.legacyTones.state.navbar : _legacyTones.legacyTones.state.default;
    var dark = stateTones.dark;
    var blend = dark ? _helpers._screen : _helpers._multiply;
    var tints = stateTones[tone] || stateTones.default;

    if (state === 'disabled') {
      return {
        bg: blend(base.bg, stateTones.default[50]),
        border: blend(base.bg, stateTones.default[50]),
        fg: blend(base.bg, stateTones.default[200])
      };
    }

    if (state === 'hovered') {
      return {
        bg: blend(base.bg, tints[50]),
        border: blend(base.bg, tints[50]),
        fg: blend(base.bg, tints[900])
      };
    }

    if (state === 'pressed') {
      return {
        bg: blend(base.bg, tints[100]),
        border: blend(base.bg, tints[100]),
        fg: blend(base.bg, tints[900])
      };
    }

    if (state === 'selected') {
      return {
        bg: blend(base.bg, tints[100]),
        border: blend(base.bg, tints[100]),
        fg: blend(base.bg, tints[900])
      };
    }

    return {
      bg: blend(base.bg, tints[100]),
      border: blend(base.bg, tints[100]),
      fg: blend(base.bg, tints[700])
    };
  },
  button: (_ref4) => {
    var base = _ref4.base,
        mode = _ref4.mode,
        muted = _ref4.muted,
        solid = _ref4.solid;

    if (mode === 'bleed') {
      return _objectSpread(_objectSpread({}, muted), {}, {
        enabled: _objectSpread(_objectSpread({}, muted.enabled), {}, {
          bg: base.bg,
          border: base.bg
        }),
        disabled: _objectSpread(_objectSpread({}, muted.disabled), {}, {
          bg: base.bg,
          border: base.bg
        })
      });
    }

    if (mode === 'ghost') {
      return _objectSpread(_objectSpread({}, solid), {}, {
        enabled: _objectSpread(_objectSpread({}, muted.enabled), {}, {
          bg: base.bg,
          border: base.border
        }),
        disabled: _objectSpread(_objectSpread({}, muted.disabled), {}, {
          bg: base.bg
        })
      });
    }

    return solid;
  },
  card: (_ref5) => {
    var base = _ref5.base,
        navbar = _ref5.dark,
        muted = _ref5.muted,
        name = _ref5.name,
        state = _ref5.state;
    var stateTones = navbar ? _legacyTones.legacyTones.state.navbar : _legacyTones.legacyTones.state.default;
    var dark = stateTones.dark;
    var blend = dark ? _helpers._screen : _helpers._multiply;

    if (state === 'selected') {
      var tint = ['default', 'transparent'].includes(name) ? stateTones.primary : stateTones[name];
      var bg = tint[500];
      var fg = stateTones.bg;
      dark = (0, _helpers._isDark)(bg, fg);
      blend = dark ? _helpers._multiply : _helpers._screen;
      return {
        bg,
        fg,
        border: tint[400],
        muted: {
          fg: blend(bg, stateTones.default[300])
        },
        accent: {
          fg: blend(bg, stateTones.critical[500])
        },
        link: {
          fg: blend(bg, stateTones.primary[300])
        },
        code: {
          bg: blend(bg, stateTones.default[950]),
          fg: blend(bg, stateTones.default[300])
        }
      };
    }

    if (state === 'hovered') {
      var _bg = muted.hovered.bg;
      return _objectSpread(_objectSpread({}, muted.hovered), {}, {
        border: blend(_bg, base.border),
        muted: {
          fg: blend(_bg, stateTones.default[700])
        },
        accent: {
          fg: blend(_bg, stateTones.critical[500])
        },
        link: {
          fg: blend(_bg, stateTones.primary[700])
        },
        code: {
          bg: blend(_bg, stateTones.default[50]),
          fg: stateTones.default[600]
        }
      });
    }

    if (state === 'pressed') {
      return _objectSpread(_objectSpread({}, muted.pressed), {}, {
        fg: base.fg,
        muted: {
          fg: blend(muted.pressed.bg, stateTones.default[700])
        },
        accent: {
          fg: blend(muted.pressed.bg, stateTones.critical[500])
        },
        link: {
          fg: blend(muted.pressed.bg, stateTones.primary[700])
        },
        code: {
          bg: blend(muted.pressed.bg, stateTones.default[50]),
          fg: stateTones.default[700]
        }
      });
    }

    if (state === 'disabled') {
      return _objectSpread(_objectSpread({}, muted.disabled), {}, {
        muted: {
          fg: muted.disabled.fg
        },
        accent: {
          fg: muted.disabled.fg
        },
        link: {
          fg: muted.disabled.fg
        },
        code: {
          bg: 'transparent',
          fg: muted.disabled.fg
        }
      });
    }

    return {
      bg: base.bg,
      fg: base.fg,
      border: base.border,
      muted: {
        fg: blend(base.bg, stateTones.default[700])
      },
      accent: {
        fg: blend(base.bg, stateTones.critical[500])
      },
      link: {
        fg: blend(base.bg, stateTones.primary[700])
      },
      code: {
        bg: blend(base.bg, stateTones.default[50]),
        fg: stateTones.default[700]
      }
    };
  },
  input: (_ref6) => {
    var base = _ref6.base,
        navbar = _ref6.dark,
        mode = _ref6.mode,
        state = _ref6.state;
    var stateTones = navbar ? _legacyTones.legacyTones.state.navbar : _legacyTones.legacyTones.state.default;
    var dark = stateTones.dark;
    var blend = dark ? _helpers._screen : _helpers._multiply;

    if (mode === 'invalid') {
      var tints = stateTones.critical;
      return {
        bg: blend(base.bg, tints[50]),
        fg: blend(base.bg, tints[700]),
        border: blend(base.bg, tints[200]),
        placeholder: blend(base.bg, tints[700])
      };
    }

    if (state === 'hovered') {
      return {
        bg: base.bg,
        fg: base.fg,
        border: blend(base.bg, _color.color.gray[300].hex),
        placeholder: blend(base.bg, _color.color.gray[700].hex)
      };
    }

    if (state === 'disabled') {
      return {
        bg: blend(base.bg, _color.color.gray[50].hex),
        fg: blend(base.bg, _color.color.gray[200].hex),
        border: blend(base.bg, _color.color.gray[100].hex),
        placeholder: blend(base.bg, _color.color.gray[100].hex)
      };
    }

    return {
      bg: base.bg,
      fg: base.fg,
      border: base.border,
      placeholder: blend(base.bg, _color.color.gray[700].hex)
    };
  },
  spot: (_ref7) => {
    var base = _ref7.base,
        key = _ref7.key;
    var dark = (0, _helpers._isDark)(base.bg, base.fg);
    var blend = dark ? _helpers._screen : _helpers._multiply;
    return blend(base.bg, _color.color[key][dark ? 400 : 500].hex);
  },
  syntax: (_ref8) => {
    var base = _ref8.base,
        navbar = _ref8.dark;
    var stateTones = navbar ? _legacyTones.legacyTones.state.navbar : _legacyTones.legacyTones.state.default;
    var dark = stateTones.dark;
    var blend = dark ? _helpers._screen : _helpers._multiply;
    var mainShade = 600;
    var secondaryShade = 400;
    return {
      atrule: blend(base.bg, _color.color.purple[mainShade].hex),
      attrName: blend(base.bg, stateTones.positive[mainShade]),
      attrValue: blend(base.bg, stateTones.caution[mainShade]),
      attribute: blend(base.bg, stateTones.caution[mainShade]),
      boolean: blend(base.bg, _color.color.purple[mainShade].hex),
      builtin: blend(base.bg, _color.color.purple[mainShade].hex),
      cdata: blend(base.bg, stateTones.caution[mainShade]),
      char: blend(base.bg, stateTones.caution[mainShade]),
      class: blend(base.bg, _color.color.orange[mainShade].hex),
      className: blend(base.bg, _color.color.cyan[mainShade].hex),
      comment: blend(base.bg, stateTones.default[secondaryShade]),
      constant: blend(base.bg, _color.color.purple[mainShade].hex),
      deleted: blend(base.bg, stateTones.critical[mainShade]),
      doctype: blend(base.bg, stateTones.default[secondaryShade]),
      entity: blend(base.bg, stateTones.critical[mainShade]),
      function: blend(base.bg, stateTones.positive[mainShade]),
      hexcode: blend(base.bg, stateTones.primary[mainShade]),
      id: blend(base.bg, _color.color.purple[mainShade].hex),
      important: blend(base.bg, _color.color.purple[mainShade].hex),
      inserted: blend(base.bg, stateTones.caution[mainShade]),
      keyword: blend(base.bg, _color.color.magenta[mainShade].hex),
      number: blend(base.bg, _color.color.purple[mainShade].hex),
      operator: blend(base.bg, _color.color.magenta[mainShade].hex),
      prolog: blend(base.bg, stateTones.default[secondaryShade]),
      property: blend(base.bg, stateTones.primary[mainShade]),
      pseudoClass: blend(base.bg, stateTones.caution[mainShade]),
      pseudoElement: blend(base.bg, stateTones.caution[mainShade]),
      punctuation: blend(base.bg, stateTones.default[mainShade]),
      regex: blend(base.bg, stateTones.primary[mainShade]),
      selector: blend(base.bg, stateTones.critical[mainShade]),
      string: blend(base.bg, stateTones.caution[mainShade]),
      symbol: blend(base.bg, _color.color.purple[mainShade].hex),
      tag: blend(base.bg, stateTones.critical[mainShade]),
      unit: blend(base.bg, _color.color.orange[mainShade].hex),
      url: blend(base.bg, stateTones.critical[mainShade]),
      variable: blend(base.bg, stateTones.critical[mainShade])
    };
  }
});
exports.color = color;
