// @sanity/form-builder/lib/inputs/DateInputs/DateTimeInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateTimeInput = void 0;

var _react = _interopRequireWildcard(require("react"));

var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));

var _CommonDateTimeInput = require("./CommonDateTimeInput");

var _legacy = require("./legacy");

var _utils = require("./utils");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var DEFAULT_DATE_FORMAT = 'YYYY-MM-DD';
var DEFAULT_TIME_FORMAT = 'HH:mm';

function parseOptions() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    dateFormat: options.dateFormat || DEFAULT_DATE_FORMAT,
    timeFormat: options.timeFormat || DEFAULT_TIME_FORMAT,
    timeStep: 'timeStep' in options && Number(options.timeStep) || 1,
    calendarTodayLabel: options.calendarTodayLabel || 'Today'
  };
}

function serialize(date) {
  return date.toISOString();
}

function deserialize(isoString) {
  var deserialized = new Date(isoString);

  if ((0, _utils.isValidDate)(deserialized)) {
    return {
      isValid: true,
      date: deserialized
    };
  }

  return {
    isValid: false,
    error: "Invalid date value: \"".concat(isoString, "\"")
  };
}

var DateTimeInput = /*#__PURE__*/_react.default.forwardRef(function DateTimeInput(props, forwardedRef) {
  var type = props.type,
      onChange = props.onChange,
      rest = _objectWithoutProperties(props, ["type", "onChange"]);

  var title = type.title,
      description = type.description,
      placeholder = type.placeholder;

  var _parseOptions = parseOptions(type.options),
      dateFormat = _parseOptions.dateFormat,
      timeFormat = _parseOptions.timeFormat,
      timeStep = _parseOptions.timeStep;

  var handleChange = (0, _react.useCallback)(nextDate => {
    onChange(_PatchEvent.default.from([nextDate === null ? (0, _PatchEvent.unset)() : (0, _PatchEvent.set)(nextDate)]));
  }, [onChange]);

  var formatInputValue = _react.default.useCallback(date => (0, _legacy.format)(date, "".concat(dateFormat, " ").concat(timeFormat)), [dateFormat, timeFormat]);

  var parseInputValue = _react.default.useCallback(inputValue => (0, _legacy.parse)(inputValue, "".concat(dateFormat, " ").concat(timeFormat)), [dateFormat, timeFormat]);

  return /*#__PURE__*/_react.default.createElement(_CommonDateTimeInput.CommonDateTimeInput, _extends({}, rest, {
    onChange: handleChange,
    ref: forwardedRef,
    selectTime: true,
    timeStep: timeStep,
    title: title,
    description: description,
    placeholder: placeholder,
    serialize: serialize,
    deserialize: deserialize,
    formatInputValue: formatInputValue,
    parseInputValue: parseInputValue
  }));
});

exports.DateTimeInput = DateTimeInput;
