// @sanity/ui/dist/es/components/autocomplete/autocomplete.js
var __defProp = Object.defineProperty;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
import {ChevronDownIcon} from "@sanity/icons";
import React, {
  cloneElement,
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import {EMPTY_ARRAY} from "../../constants";
import {focusFirstDescendant} from "../../helpers";
import {useForwardedRef, useResponsiveProp} from "../../hooks";
import {Box, Button, Card, Spinner, Text, TextInput} from "../../primitives";
import {AutocompleteOption} from "./autocompleteOption";
import {Root, LoadingCard, ListBoxContainer, ListBoxCard} from "./styles";
const EMPTY_RECORD = {};
const defaultRenderValue = (value, option) => option ? option.value : value;
const defaultFilterOption = (query, option) => option.value.toLowerCase().indexOf(query.toLowerCase()) > -1;
const LIST_IGNORE_KEYS = [
  "Control",
  "Shift",
  "Alt",
  "Enter",
  "Home",
  "End",
  "PageUp",
  "PageDown",
  "Meta",
  "Tab",
  "CapsLock"
];
const InnerAutocomplete = forwardRef((props, ref) => {
  const {
    border = true,
    filterOption: filterOptionProp,
    fontSize = 2,
    icon,
    id,
    loading,
    onChange,
    onQueryChange,
    onSelect,
    openButton,
    options: optionsProp,
    padding: paddingProp = 3,
    radius = 2,
    renderOption: renderOptionProp,
    renderValue = defaultRenderValue,
    value: valueProp = ""
  } = props, restProps = __rest(props, [
    "border",
    "filterOption",
    "fontSize",
    "icon",
    "id",
    "loading",
    "onChange",
    "onQueryChange",
    "onSelect",
    "openButton",
    "options",
    "padding",
    "radius",
    "renderOption",
    "renderValue",
    "value"
  ]);
  const defaultRenderOption = useCallback(({value: value2}) => /* @__PURE__ */ React.createElement(Card, {
    as: "button",
    padding: paddingProp,
    tone: "inherit"
  }, /* @__PURE__ */ React.createElement(Text, {
    size: fontSize
  }, value2)), [fontSize, paddingProp]);
  const renderOption = typeof renderOptionProp === "function" ? renderOptionProp : defaultRenderOption;
  const filterOption = typeof filterOptionProp === "function" ? filterOptionProp : defaultFilterOption;
  const [value, setValue] = useState(valueProp);
  const [query, setQuery] = useState(null);
  const valueRef = useRef(value);
  const [focused, setFocused] = useState(false);
  const listboxId = `${id}-listbox`;
  const options = Array.isArray(optionsProp) ? optionsProp : EMPTY_ARRAY;
  const [selectedIndex, setSelectedIndex] = useState(-1);
  const inputRef = useRef(null);
  const listRef = useRef(null);
  const activeItemId = selectedIndex > -1 ? `${id}-option-${selectedIndex}` : void 0;
  const padding = useResponsiveProp(paddingProp);
  const rootRef = useRef(null);
  const currentOption = value ? options.find((o) => o.value === value) : void 0;
  const filteredOptions = useMemo(() => options.filter((option) => query ? filterOption(query, option) : true), [filterOption, options, query]);
  const optionsLen = filteredOptions.length;
  const expanded = loading || focused && optionsLen > 0 && query !== null;
  const forwardedRef = useForwardedRef(ref);
  const handleRootBlur = useCallback(() => {
    setTimeout(() => {
      const rootEl = rootRef.current;
      const focusedEl = document.activeElement;
      const focusInside = rootEl && focusedEl && rootEl.contains(focusedEl);
      if (!focusInside)
        setFocused(false);
    }, 0);
  }, []);
  const handleRootKeyDown = useCallback((event) => {
    var _a, _b;
    if (event.key === "ArrowDown") {
      event.preventDefault();
      if (!optionsLen)
        return;
      setSelectedIndex((index) => (index + 1) % optionsLen);
      return;
    }
    if (event.key === "ArrowUp") {
      event.preventDefault();
      if (!optionsLen)
        return;
      setSelectedIndex((index) => {
        return index === -1 ? optionsLen - 1 : (optionsLen + index - 1) % optionsLen;
      });
      return;
    }
    if (event.key === "Escape") {
      setFocused(false);
      valueRef.current = "";
      setQuery(null);
      if (onQueryChange)
        onQueryChange(null);
      (_a = inputRef.current) == null ? void 0 : _a.focus();
      return;
    }
    const target = event.target;
    const listEl = listRef.current;
    if ((listEl === target || (listEl == null ? void 0 : listEl.contains(target))) && !LIST_IGNORE_KEYS.includes(event.key)) {
      (_b = inputRef.current) == null ? void 0 : _b.focus();
      return;
    }
  }, [onQueryChange, optionsLen]);
  const handleInputChange = useCallback((event) => {
    const q = event.currentTarget.value;
    valueRef.current = q;
    setQuery(q);
    if (onQueryChange)
      onQueryChange(q);
  }, [onQueryChange]);
  const handleInputFocus = useCallback(() => setFocused(true), []);
  const handleClearButtonClick = useCallback(() => {
    var _a;
    valueRef.current = "";
    setValue("");
    if (onChange)
      onChange("");
    setQuery(null);
    if (onQueryChange)
      onQueryChange(null);
    (_a = inputRef.current) == null ? void 0 : _a.focus();
  }, [onChange, onQueryChange]);
  const handleClearButtonFocus = useCallback(() => setFocused(true), []);
  const handleOptionSelect = useCallback((v) => {
    var _a;
    if (onSelect)
      onSelect(v);
    setValue(v);
    if (onChange)
      onChange(v);
    setQuery(null);
    if (onQueryChange)
      onQueryChange(null);
    setFocused(false);
    (_a = inputRef.current) == null ? void 0 : _a.focus();
  }, [onChange, onSelect, onQueryChange]);
  const handleOpenClick = useCallback(() => {
    var _a;
    (_a = inputRef.current) == null ? void 0 : _a.focus();
    setQuery(query || "");
  }, [query]);
  useEffect(() => {
    if (valueProp !== valueRef.current) {
      valueRef.current = valueProp;
      setValue(valueProp);
    }
  }, [valueProp]);
  useEffect(() => setSelectedIndex(-1), [optionsLen]);
  useEffect(() => {
    const listElement = listRef.current;
    if (!listElement)
      return;
    const selectedItemElement = listElement.childNodes[selectedIndex];
    if (selectedItemElement) {
      focusFirstDescendant(selectedItemElement);
    }
  }, [selectedIndex]);
  const setRef = (el) => {
    inputRef.current = el;
    forwardedRef.current = el;
  };
  const clearButton = useMemo(() => value.length > 0 ? {
    "aria-label": "Clear",
    onFocus: handleClearButtonFocus
  } : void 0, [handleClearButtonFocus, value]);
  const openButtonBoxPadding = useMemo(() => padding.map((v) => v - 2), [padding]);
  const openButtonPadding = useMemo(() => padding.map((v) => v - 1), [padding]);
  const openButtonProps = useMemo(() => typeof openButton === "object" ? openButton : EMPTY_RECORD, [openButton]);
  const openButtonNode = useMemo(() => openButton ? /* @__PURE__ */ React.createElement(Box, {
    padding: openButtonBoxPadding
  }, /* @__PURE__ */ React.createElement(Button, __assign(__assign({}, openButtonProps), {
    fontSize,
    icon: ChevronDownIcon,
    mode: "bleed",
    onClick: handleOpenClick,
    padding: openButtonPadding
  }))) : void 0, [
    fontSize,
    handleOpenClick,
    openButton,
    openButtonBoxPadding,
    openButtonPadding,
    openButtonProps
  ]);
  return /* @__PURE__ */ React.createElement(Root, {
    "data-ui": "Autocomplete",
    onBlur: handleRootBlur,
    onKeyDown: handleRootKeyDown,
    ref: rootRef
  }, /* @__PURE__ */ React.createElement(TextInput, __assign(__assign({}, restProps), {
    "aria-activedescendant": activeItemId,
    "aria-autocomplete": "list",
    "aria-expanded": expanded,
    "aria-owns": listboxId,
    autoCapitalize: "off",
    autoComplete: "off",
    autoCorrect: "off",
    border,
    clearButton,
    fontSize,
    icon,
    id,
    inputMode: "search",
    onChange: handleInputChange,
    onClear: handleClearButtonClick,
    onFocus: handleInputFocus,
    padding,
    radius,
    ref: setRef,
    role: "combobox",
    spellCheck: false,
    suffix: openButtonNode,
    value: query === null ? renderValue(value, currentOption) : query
  })), /* @__PURE__ */ React.createElement(ListBoxContainer, {
    hidden: !expanded
  }, /* @__PURE__ */ React.createElement(ListBoxCard, {
    paddingY: 1,
    radius: 1,
    shadow: 2,
    tabIndex: -1,
    tone: "inherit"
  }, /* @__PURE__ */ React.createElement("ul", {
    "aria-multiselectable": false,
    id: listboxId,
    ref: listRef,
    role: "listbox"
  }, filteredOptions.map((option, optionIndex) => /* @__PURE__ */ React.createElement(AutocompleteOption, {
    id: `${id}-option-${optionIndex}`,
    key: optionIndex,
    onSelect: handleOptionSelect,
    selected: selectedIndex > -1 ? optionIndex === selectedIndex : currentOption === option,
    value: option.value
  }, cloneElement(renderOption(option), {disabled: loading, tabIndex: -1})))), /* @__PURE__ */ React.createElement(LoadingCard, {
    padding: 3,
    style: {opacity: loading ? 0.5 : 0}
  }, /* @__PURE__ */ React.createElement(Spinner, null)))));
});
InnerAutocomplete.displayName = "Autocomplete";
const Autocomplete = InnerAutocomplete;
export {
  Autocomplete
};
