// @sanity/ui/dist/es/primitives/flex/flex.js
var __defProp = Object.defineProperty;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
import React, {forwardRef} from "react";
import styled from "styled-components";
import {
  flexItemStyle,
  responsiveFlexStyle
} from "../../styles/internal";
import {Box} from "../box";
const Root = styled(Box)(flexItemStyle, responsiveFlexStyle);
const Flex = forwardRef((props, ref) => {
  const {align, as, direction = "row", justify, wrap} = props, restProps = __rest(props, ["align", "as", "direction", "justify", "wrap"]);
  return /* @__PURE__ */ React.createElement(Root, __assign(__assign({
    "data-ui": "Flex"
  }, restProps), {
    $align: align,
    $direction: direction,
    $justify: justify,
    $wrap: wrap,
    forwardedAs: as,
    ref
  }));
});
Flex.displayName = "Flex";
export {
  Flex
};
