// @sanity/form-builder/lib/sanity/inputs/SanityImageInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ImageInput = _interopRequireDefault(require("../../inputs/files/ImageInput"));

var _resolveUploader = _interopRequireDefault(require("../uploads/resolveUploader"));

var _withDocument = _interopRequireDefault(require("../../utils/withDocument"));

var _legacyParts = require("../../legacyParts");

var _assets = require("./client-adapters/assets");

var _formBuilderConfig$im;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var globalAssetSources = _legacyParts.userDefinedAssetSources ? _legacyParts.userDefinedAssetSources : _legacyParts.defaultAssetSources;
var SUPPORT_DIRECT_UPLOADS = (_legacyParts.formBuilderConfig === null || _legacyParts.formBuilderConfig === void 0 ? void 0 : (_formBuilderConfig$im = _legacyParts.formBuilderConfig.images) === null || _formBuilderConfig$im === void 0 ? void 0 : _formBuilderConfig$im.directUploads) !== false;

var _default = /*#__PURE__*/_react.default.forwardRef(function SanityImageInput(props, forwardedRef) {
  var _props$type$options;

  var sourcesFromSchema = (_props$type$options = props.type.options) === null || _props$type$options === void 0 ? void 0 : _props$type$options.sources; // note: type.options.sources may be an empty array and in that case we're
  // disabling selecting images from asset source  (it's a feature, not a bug)

  var assetSources = _react.default.useMemo(() => (sourcesFromSchema || globalAssetSources).map(source => _objectSpread(_objectSpread({}, source), {}, {
    // Note: The asset source plugin get's passed the enclosing document by default.
    // This is a potential performance hog, so we should consider some alternatives here
    // e.g. we could offer a way to declare the path for the values in the document you're interested in instead
    component: (0, _withDocument.default)(source.component)
  })), [sourcesFromSchema]);

  return /*#__PURE__*/_react.default.createElement(_ImageInput.default, _extends({}, props, {
    resolveUploader: _resolveUploader.default,
    materialize: _assets.materializeReference,
    assetSources: assetSources,
    directUploads: SUPPORT_DIRECT_UPLOADS,
    ref: forwardedRef
  }));
});

exports.default = _default;
