// @sanity/ui/dist/cjs/primitives/checkbox/styles.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  checkboxBaseStyles: () => checkboxBaseStyles,
  inputElementStyles: () => inputElementStyles
});
var import_styled_components = __toModule(require("styled-components"));
var import_styles = __toModule(require("../../styles"));
var import_internal = __toModule(require("../../styles/internal"));
function checkboxBaseStyles() {
  return import_styled_components.css`
    position: relative;
    display: inline-block;
  `;
}
function inputElementStyles(props) {
  const {theme} = props;
  const color = theme.sanity.color.input;
  const {focusRing, input, radius} = theme.sanity;
  return import_styled_components.css`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    outline: none;
    opacity: 0;
    z-index: 1;
    padding: 0;
    margin: 0;

    & + span {
      position: relative;
      display: block;
      height: ${(0, import_styles.rem)(input.checkbox.size)};
      width: ${(0, import_styles.rem)(input.checkbox.size)};
      box-sizing: border-box;
      box-shadow: ${(0, import_internal.focusRingBorderStyle)({
    color: color.default.enabled.border,
    width: input.border.width
  })};
      border-radius: ${(0, import_styles.rem)(radius[2])};
      line-height: 1;
      background: ${color.default.enabled.bg};

      & > svg {
        display: block;
        position: absolute;
        opacity: 0;
        height: 100%;
        width: 100%;

        & > path {
          vector-effect: non-scaling-stroke;
          stroke-width: 2 !important;
        }
      }
    }

    &:not(:disabled):focus + span {
      box-shadow: ${(0, import_internal.focusRingStyle)({
    border: {width: input.border.width, color: color.default.enabled.border},
    focusRing
  })};
    }

    &:not(:disabled):focus:not(:focus-visible) + span {
      box-shadow: ${(0, import_internal.focusRingBorderStyle)({
    color: color.default.enabled.border,
    width: input.border.width
  })};
    }

    &:checked + span > svg:first-child {
      opacity: 1;
    }

    &:not([data-read-only]):disabled + span {
      background: ${color.default.disabled.bg};
      box-shadow: ${(0, import_internal.focusRingBorderStyle)({
    width: input.border.width,
    color: color.default.disabled.border
  })};
      color: ${color.default.disabled.fg};
    }

    &:indeterminate + span > svg:last-child {
      opacity: 1;
    }
  `;
}
