// @sanity/ui/dist/es/primitives/card/card.js
var __defProp = Object.defineProperty;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
import React, {forwardRef} from "react";
import {isValidElementType} from "react-is";
import styled from "styled-components";
import {
  responsiveBorderStyle,
  responsiveRadiusStyle,
  responsiveShadowStyle
} from "../../styles/internal";
import {ThemeColorProvider, useRootTheme} from "../../theme";
import {Box} from "../box";
import {cardStyle} from "./styles";
const Root = styled(Box)(responsiveBorderStyle, responsiveRadiusStyle, responsiveShadowStyle, cardStyle);
const Card = forwardRef((props, ref) => {
  const {
    as: asProp,
    border,
    borderTop,
    borderRight,
    borderBottom,
    borderLeft,
    radius = 0,
    scheme,
    shadow,
    tone: toneProp = "default"
  } = props, restProps = __rest(props, [
    "as",
    "border",
    "borderTop",
    "borderRight",
    "borderBottom",
    "borderLeft",
    "radius",
    "scheme",
    "shadow",
    "tone"
  ]);
  const as = isValidElementType(asProp) ? asProp : "div";
  const rootTheme = useRootTheme();
  const tone = toneProp === "inherit" ? rootTheme.tone : toneProp;
  return /* @__PURE__ */ React.createElement(ThemeColorProvider, {
    scheme,
    tone
  }, /* @__PURE__ */ React.createElement(Root, __assign(__assign({
    "data-as": typeof as === "string" ? as : void 0,
    "data-scheme": rootTheme.scheme,
    "data-ui": "Card",
    "data-tone": tone
  }, restProps), {
    $border: border,
    $borderTop: borderTop,
    $borderRight: borderRight,
    $borderBottom: borderBottom,
    $borderLeft: borderLeft,
    $radius: radius,
    $shadow: shadow,
    $tone: tone,
    forwardedAs: as,
    ref
  })));
});
Card.displayName = "Card";
export {
  Card
};
