// @sanity/dashboard/lib/widgets/projectUsers/ProjectUsers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _operators = require("rxjs/operators");

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/previews/default"));

var _default3 = require("part:@sanity/components/lists/default");

var _anchor = _interopRequireDefault(require("part:@sanity/components/buttons/anchor"));

var _robotIcon = _interopRequireDefault(require("part:@sanity/base/robot-icon"));

var _versionedClient = require("../../versionedClient");

var _ProjectUsers = _interopRequireDefault(require("./ProjectUsers.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function getInviteUrl(projectId) {
  return "https://manage.sanity.io/projects/".concat(projectId, "/team/invite");
}

function sortUsersByRobotStatus(userA, userB, project) {
  var members = project.members;
  var membershipA = members.find(member => member.id === userA.id);
  var membershipB = members.find(member => member.id === userB.id);

  if (membershipA.isRobot) {
    return 1;
  }

  if (membershipB.isRobot) {
    return -1;
  }

  return 0;
}

class ProjectUsers extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      project: null,
      users: null,
      error: null
    });

    _defineProperty(this, "handleRetryFetch", () => {
      this.fetchData();
    });
  }

  componentDidMount() {
    this.fetchData();
  }

  componentWillUnmount() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  fetchData() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }

    var _versionedClient$conf = _versionedClient.versionedClient.config(),
        projectId = _versionedClient$conf.projectId;

    this.subscription = _versionedClient.versionedClient.observable.request({
      uri: "/projects/".concat(projectId)
    }).pipe((0, _operators.switchMap)(project => _versionedClient.versionedClient.observable.request({
      uri: "/users/".concat(project.members.map(mem => mem.id).join(','))
    }).pipe((0, _operators.map)(users => ({
      project,
      users
    }))))).subscribe({
      next: (_ref) => {
        var users = _ref.users,
            project = _ref.project;
        return this.setState({
          project,
          users: (Array.isArray(users) ? users : [users]).sort((userA, userB) => sortUsersByRobotStatus(userA, userB, project))
        });
      },
      error: _error => this.setState({
        error: _error
      })
    });
  }

  render() {
    var _this$state = this.state,
        error = _this$state.error,
        project = _this$state.project,
        users = _this$state.users;
    var isLoading = !users || !project;

    if (error) {
      return /*#__PURE__*/_react.default.createElement("div", null, "Something went wrong while fetching data. You could", ' ', /*#__PURE__*/_react.default.createElement("a", {
        className: _ProjectUsers.default.retry,
        onClick: this.handleRetryFetch,
        title: "Retry users fetch"
      }, "retry"), "..?");
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: _ProjectUsers.default.container
    }, /*#__PURE__*/_react.default.createElement("header", {
      className: _ProjectUsers.default.header
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _ProjectUsers.default.title
    }, "Project users")), isLoading && /*#__PURE__*/_react.default.createElement(_default3.List, {
      className: _ProjectUsers.default.list
    }, /*#__PURE__*/_react.default.createElement(_spinner.default, {
      center: true,
      message: "Loading items\u2026"
    })), !isLoading && /*#__PURE__*/_react.default.createElement(_default3.List, {
      className: _ProjectUsers.default.list
    }, users.map(user => {
      var membership = project.members.find(member => member.id === user.id);
      var media = membership.isRobot ? /*#__PURE__*/_react.default.createElement(_robotIcon.default, {
        className: _ProjectUsers.default.profileImage
      }) : /*#__PURE__*/_react.default.createElement("div", {
        className: _ProjectUsers.default.avatar
      }, user.imageUrl && /*#__PURE__*/_react.default.createElement("img", {
        src: user.imageUrl,
        alt: user.displayName
      }));
      return /*#__PURE__*/_react.default.createElement(_default3.Item, {
        key: user.id,
        className: _ProjectUsers.default.item
      }, /*#__PURE__*/_react.default.createElement(_default2.default, {
        title: user.displayName,
        subtitle: membership.role,
        media: media
      }));
    })), /*#__PURE__*/_react.default.createElement("div", {
      className: _ProjectUsers.default.footer
    }, /*#__PURE__*/_react.default.createElement(_anchor.default, {
      disabled: isLoading,
      href: isLoading ? undefined : getInviteUrl(project.id),
      bleed: true,
      color: "primary",
      kind: "simple"
    }, "Invite members")));
  }

}

_defineProperty(ProjectUsers, "propTypes", {});

_defineProperty(ProjectUsers, "defaultProps", {});

var _default = ProjectUsers;
exports.default = _default;
