// @sanity/ui/dist/cjs/primitives/card/styles.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  cardBaseStyle: () => cardBaseStyle,
  cardColorStyle: () => cardColorStyle,
  cardStyle: () => cardStyle
});
var import_styled_components = __toModule(require("styled-components"));
function cardStyle(props) {
  return [cardBaseStyle, cardColorStyle(props)];
}
function cardBaseStyle() {
  return import_styled_components.css`
    &[data-as='button'] {
      -webkit-font-smoothing: inherit;
      appearance: none;
      outline: none;
      font: inherit;
      text-align: inherit;
      border: 0;
      width: stretch;
    }

    /* &:is(a) */
    &[data-as='a'] {
      text-decoration: none;
    }

    /* &:is(pre) */
    &[data-as='pre'] {
      font: inherit;
    }
  `;
}
function vars(base, color) {
  return import_styled_components.css`
    --card-bg-color: ${color.bg};
    --card-fg-color: ${color.fg};
    --card-focus-ring-color: ${base.focusRing};
    --card-border-color: ${color.border};
    --card-muted-fg-color: ${color.muted.fg};
    --card-accent-fg-color: ${color.accent.fg};
    --card-link-fg-color: ${color.link.fg};
    --card-code-bg-color: ${color.code.bg};
    --card-code-fg-color: ${color.code.fg};

    /* @todo: deprecate */
    --card-link-color: ${color.link.fg};
    --card-hairline-soft-color: ${color.border};
    --card-hairline-hard-color: ${color.border};

    /* @todo: rename to "--base-"? */
    --card-shadow-outline-color: ${base.shadow.outline};
    --card-shadow-umbra-color: ${base.shadow.umbra};
    --card-shadow-penumbra-color: ${base.shadow.penumbra};
    --card-shadow-ambient-color: ${base.shadow.ambient};
  `;
}
function cardColorStyle(props) {
  var _a, _b;
  const {theme} = props;
  const {base, card} = theme.sanity.color;
  return import_styled_components.css`
    ${vars(base, card.enabled)}

    background-color: var(--card-bg-color);
    color: var(--card-fg-color);

    /* &:is(button) */
    &[data-as='button'] {
      &:disabled {
        ${vars(base, card.disabled)}
      }

      &:not(:disabled) {
        @media (hover: hover) {
          &:hover {
            ${vars(base, card.hovered)}
          }

          &:active {
            ${vars(base, card.pressed)}
          }
        }

        &:focus {
          ${vars(base, card.selected)}
        }

        &[aria-pressed='true'],
        [aria-selected='true'] > & {
          ${vars(base, card.selected)}
        }
      }
    }

    /* &:is(a) */
    &[data-as='a'] {
      &[data-disabled] {
        ${vars(base, card.disabled)}
      }

      &:not([data-disabled]) {
        @media (hover: hover) {
          outline: none;

          &:hover {
            ${vars(base, card.hovered)}
          }

          &:active {
            ${vars(base, card.pressed)}
          }
        }

        &:focus {
          ${vars(base, card.selected)}
        }

        [aria-selected='true'] > & {
          ${vars(base, card.selected)}
        }
      }
    }

    ${(_b = (_a = theme.sanity.styles) == null ? void 0 : _a.card) == null ? void 0 : _b.root}
  `;
}
