// @sanity/dashboard/lib/widgets/sanityTutorials/Tutorial.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Tutorial = _interopRequireDefault(require("./Tutorial.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Tutorial extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        title = _this$props.title,
        posterURL = _this$props.posterURL,
        showPlayIcon = _this$props.showPlayIcon,
        href = _this$props.href,
        presenterName = _this$props.presenterName,
        presenterSubtitle = _this$props.presenterSubtitle;
    return /*#__PURE__*/_react.default.createElement("a", {
      className: _Tutorial.default.root,
      href: href,
      target: "_blank",
      rel: "noopener noreferrer"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _Tutorial.default.posterContainer
    }, /*#__PURE__*/_react.default.createElement("img", {
      className: _Tutorial.default.poster,
      src: posterURL
    }), showPlayIcon && /*#__PURE__*/_react.default.createElement("div", {
      className: _Tutorial.default.playIcon
    }, /*#__PURE__*/_react.default.createElement("span", null))), /*#__PURE__*/_react.default.createElement("h3", {
      className: _Tutorial.default.title
    }, title), /*#__PURE__*/_react.default.createElement("div", {
      className: _Tutorial.default.byLine
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _Tutorial.default.presenterName
    }, presenterName), /*#__PURE__*/_react.default.createElement("div", {
      className: _Tutorial.default.presenterSubtitle
    }, presenterSubtitle)));
  }

}

_defineProperty(Tutorial, "propTypes", {
  title: _propTypes.default.string.isRequired,
  posterURL: _propTypes.default.string,
  href: _propTypes.default.string.isRequired,
  showPlayIcon: _propTypes.default.bool,
  presenterName: _propTypes.default.string.isRequired,
  presenterSubtitle: _propTypes.default.string.isRequired
});

_defineProperty(Tutorial, "defaultProps", {
  posterURL: null,
  showPlayIcon: false
});

var _default = Tutorial;
exports.default = _default;
