// @sanity/studio-hints/lib/components/HintCard.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _launchIcon = _interopRequireDefault(require("part:@sanity/base/launch-icon"));

var _HintCard = _interopRequireDefault(require("./HintCard.css"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function HintCard(props) {
  var card = props.card,
      repoId = props.repoId;
  /**
   * Hint cards currently only link to external sources.
   * In future iterations a hint card may have it's own page
   * that opens by clicking the card (onCardClick prop), that
   * then renders a page component.
   */

  /*
  <div className={styles.root} onClick={() => onCardClick(card._id)}>
    <h4 className={styles.cardTitle}>
      {card.title} <ArrowRight />
    </h4>
    <p className={styles.cardSummary}>{card.summary}</p>
  </div>
  */

  return /*#__PURE__*/_react.default.createElement("li", null, /*#__PURE__*/_react.default.createElement("a", {
    href: (0, _utils.resolveUrl)(card.hint, repoId),
    className: _HintCard.default.root,
    target: "_blank",
    rel: "noopener noreferrer"
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("span", {
    className: _HintCard.default.cardTitle
  }, card.titleOverride || card.hint.title), /*#__PURE__*/_react.default.createElement("p", {
    className: _HintCard.default.cardSummary
  }, card.hint.description)), /*#__PURE__*/_react.default.createElement("span", {
    className: _HintCard.default.externalIcon
  }, /*#__PURE__*/_react.default.createElement(_launchIcon.default, null))));
}

HintCard.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  card: _propTypes.default.object.isRequired,
  repoId: _propTypes.default.string.isRequired // onCardClick: PropTypes.func.isRequired

};
var _default = HintCard;
exports.default = _default;
