// @sanity/studio-hints/lib/components/LinksList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _launchIcon = _interopRequireDefault(require("part:@sanity/base/launch-icon"));

var _LinksList = _interopRequireDefault(require("./LinksList.css"));

var _HintCard = _interopRequireDefault(require("./HintCard"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function CardLinks(props) {
  var type = props.type,
      links = props.links,
      title = props.title,
      repoId = props.repoId;

  if (!links) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: _LinksList.default.root
  }, /*#__PURE__*/_react.default.createElement("h3", {
    className: _LinksList.default.listTitle
  }, title), /*#__PURE__*/_react.default.createElement("ul", {
    className: "".concat(_LinksList.default.linksList, " ").concat(type === 'card' ? _LinksList.default.cardList : _LinksList.default.simpleList)
  }, links.map(link => {
    return type === 'card' ? /*#__PURE__*/_react.default.createElement(_HintCard.default, {
      key: link._key,
      card: link,
      repoId: repoId
    }) : /*#__PURE__*/_react.default.createElement("li", {
      className: _LinksList.default.linkTitle,
      key: link.title
    }, /*#__PURE__*/_react.default.createElement("a", {
      className: _LinksList.default.link,
      href: "".concat(link.url, "?utm_source=hints&utm_medium=").concat(repoId),
      target: "_blank",
      rel: "noopener noreferrer"
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: _LinksList.default.linkTitleText
    }, link.title), /*#__PURE__*/_react.default.createElement("span", {
      className: _LinksList.default.externalIcon
    }, /*#__PURE__*/_react.default.createElement(_launchIcon.default, null))));
  })));
}

CardLinks.propTypes = {
  type: _propTypes.default.string,
  title: _propTypes.default.string,
  repoId: _propTypes.default.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  links: _propTypes.default.array.isRequired
};
CardLinks.defaultProps = {
  type: null,
  title: ''
};
var _default = CardLinks;
exports.default = _default;
