// @sanity/studio-hints/lib/components/ToggleSidecarButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _sidecarDatastore = require("part:@sanity/default-layout/sidecar-datastore");

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _helpCircleIcon = _interopRequireDefault(require("part:@sanity/base/help-circle-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ToggleSidecarButton extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      isOpen: true
    });
  }

  componentDidMount() {
    this.subscription = _sidecarDatastore.isSidecarOpenSetting.listen().subscribe(isOpen => {
      this.setState({
        isOpen: isOpen !== false
      });
    });
  }

  componentWillUnmount() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  render() {
    var isOpen = this.state.isOpen;
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      "aria-label": "Toggle sidecar",
      "aria-pressed": isOpen,
      icon: _helpCircleIcon.default,
      kind: "simple",
      onClick: _sidecarDatastore.toggleSidecarOpenState,
      selected: isOpen,
      tone: "navbar",
      padding: "small"
    });
  }

}

exports.default = ToggleSidecarButton;
