// @sanity/form-builder/lib/sanity/inputResolver/resolveArrayInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isArrayOfPrimitives = isArrayOfPrimitives;
exports.hasOptionsList = hasOptionsList;
exports.default = resolveArrayInput;

var is = _interopRequireWildcard(require("../../utils/is"));

var _OptionsArrayInput = _interopRequireDefault(require("../../inputs/OptionsArrayInput"));

var _PortableTextInput = _interopRequireDefault(require("../../inputs/PortableText/PortableTextInput"));

var _ArrayOfPrimitivesInput = _interopRequireDefault(require("../../inputs/ArrayOfPrimitivesInput"));

var _TagsArrayInput = _interopRequireDefault(require("../../inputs/TagsArrayInput"));

var _SanityArrayInput = _interopRequireDefault(require("../inputs/SanityArrayInput"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var PRIMITIVES = ['string', 'number', 'boolean'];

function isArrayOfPrimitives(type) {
  return type.of.every(ofType => PRIMITIVES.includes(ofType.jsonType));
}

function isStringArray(type) {
  return type.of.length === 1 && is.type('string', type.of[0]);
}

function isTagsArray(type) {
  var _type$options;

  return ((_type$options = type.options) === null || _type$options === void 0 ? void 0 : _type$options.layout) === 'tags';
}

function isPortableText(type) {
  // TODO: better testing here, not only for type 'block' !
  return type.of.some(memberType => is.type('block', memberType));
}

function hasOptionsList(type) {
  var _type$options2;

  return Boolean((_type$options2 = type.options) === null || _type$options2 === void 0 ? void 0 : _type$options2.list);
}

function resolveArrayInput(type) {
  // Schema provides predefines list
  if (hasOptionsList(type)) {
    return _OptionsArrayInput.default;
  }

  if (isStringArray(type) && isTagsArray(type)) {
    return _TagsArrayInput.default;
  } // Special component for array of primitive values


  if (isArrayOfPrimitives(type)) {
    return _ArrayOfPrimitivesInput.default;
  } // Use Portable Text editor if portable text.


  if (isPortableText(type)) {
    return _PortableTextInput.default;
  } // use default


  return _SanityArrayInput.default;
}
