// @sanity/ui/dist/es/styles/flex/flexStyle.js
import {getResponsiveProp, responsive} from "../helpers";
function responsiveFlexStyle() {
  return [
    flexBaseStyle,
    responsiveFlexAlignStyle,
    responsiveFlexWrapStyle,
    responsiveFlexJustifyStyle,
    responsiveFlexDirectionStyle
  ];
}
function flexBaseStyle() {
  return {"&:not([hidden])": {display: "flex"}};
}
function responsiveFlexAlignStyle(props) {
  const {theme} = props;
  const {media} = theme.sanity;
  return responsive(media, getResponsiveProp(props.$align), (align) => {
    return {alignItems: align};
  });
}
function responsiveFlexWrapStyle(props) {
  const {theme} = props;
  const {media} = theme.sanity;
  return responsive(media, getResponsiveProp(props.$wrap), (wrap) => {
    return {flexWrap: wrap};
  });
}
function responsiveFlexJustifyStyle(props) {
  const {theme} = props;
  const {media} = theme.sanity;
  return responsive(media, getResponsiveProp(props.$justify), (justify) => {
    return {justifyContent: justify};
  });
}
function responsiveFlexDirectionStyle(props) {
  const {theme} = props;
  const {media} = theme.sanity;
  return responsive(media, getResponsiveProp(props.$direction), (direction) => {
    return {flexDirection: direction};
  });
}
export {
  responsiveFlexAlignStyle,
  responsiveFlexDirectionStyle,
  responsiveFlexJustifyStyle,
  responsiveFlexStyle,
  responsiveFlexWrapStyle
};
