// @sanity/ui/dist/es/primitives/tooltip/tooltip.js
var __defProp = Object.defineProperty;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
import React, {cloneElement, forwardRef, useCallback, useEffect, useState} from "react";
import {usePopper} from "react-popper";
import styled from "styled-components";
import {useForwardedRef} from "../../hooks";
import {Layer, Portal, useBoundaryElement} from "../../utils";
import {Card} from "../card";
import {TooltipArrow} from "./tooltipArrow";
const Root = styled(Layer)`
  pointer-events: none;
`;
const Tooltip = forwardRef((props, ref) => {
  const boundaryElementContext = useBoundaryElement();
  const {
    allowedAutoPlacements,
    boundaryElement = boundaryElementContext == null ? void 0 : boundaryElementContext.element,
    children,
    content,
    disabled,
    fallbackPlacements,
    placement = "bottom",
    portal
  } = props, restProps = __rest(props, [
    "allowedAutoPlacements",
    "boundaryElement",
    "children",
    "content",
    "disabled",
    "fallbackPlacements",
    "placement",
    "portal"
  ]);
  const forwardedRef = useForwardedRef(ref);
  const [referenceElement, setReferenceElement] = useState(null);
  const [popperElement, setPopperElement] = useState(null);
  const [arrowElement, setArrowElement] = useState(null);
  const popper = usePopper(referenceElement, popperElement, {
    placement,
    modifiers: [
      {
        name: "arrow",
        options: {
          element: arrowElement,
          padding: 4
        }
      },
      {
        name: "preventOverflow",
        options: {
          altAxis: true,
          boundary: boundaryElement || void 0,
          padding: 4
        }
      },
      {
        name: "offset",
        options: {offset: [0, 3]}
      },
      {
        name: "flip",
        options: {
          allowedAutoPlacements,
          fallbackPlacements
        }
      }
    ]
  });
  const {forceUpdate} = popper;
  const [isOpen, setIsOpen] = useState(false);
  const handleBlur = useCallback(() => setIsOpen(false), []);
  const handleFocus = useCallback(() => setIsOpen(true), []);
  const handleMouseEnter = useCallback(() => setIsOpen(true), []);
  const handleMouseLeave = useCallback(() => setIsOpen(false), []);
  useEffect(() => {
    if (forceUpdate)
      forceUpdate();
  }, [forceUpdate, content]);
  const setRef = (el) => {
    setPopperElement(el);
    forwardedRef.current = el;
  };
  if (!children)
    return /* @__PURE__ */ React.createElement(React.Fragment, null);
  if (disabled)
    return children;
  const referenceProps = {
    onBlur: handleBlur,
    onFocus: handleFocus,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    ref: setReferenceElement
  };
  const referenceNode = cloneElement(children, referenceProps);
  const popperNode = /* @__PURE__ */ React.createElement(Root, __assign(__assign(__assign({
    "data-ui": "Tooltip"
  }, restProps), {
    ref: setRef,
    style: popper.styles.popper
  }), popper.attributes.popper), /* @__PURE__ */ React.createElement(Card, {
    radius: 2,
    shadow: 3
  }, content, /* @__PURE__ */ React.createElement(TooltipArrow, {
    ref: setArrowElement,
    style: popper.styles.arrow
  })));
  return /* @__PURE__ */ React.createElement(React.Fragment, null, referenceNode, isOpen && /* @__PURE__ */ React.createElement(React.Fragment, null, portal && /* @__PURE__ */ React.createElement(Portal, null, popperNode), !portal && popperNode));
});
Tooltip.displayName = "Tooltip";
export {
  Tooltip
};
