// @sanity/ui/dist/es/primitives/text/text.js
var __defProp = Object.defineProperty;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
import React, {forwardRef} from "react";
import styled from "styled-components";
import {
  responsiveTextAlignStyle,
  responsiveTextFont
} from "../../styles/internal";
import {textBaseStyle} from "./styles";
const Root = styled.div(responsiveTextFont, responsiveTextAlignStyle, textBaseStyle);
const SpanWithTextOverflow = styled.span`
  display: block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;
const Text = forwardRef((props, ref) => {
  const {
    accent = false,
    align,
    children: childrenProp,
    muted = false,
    size = 2,
    textOverflow,
    weight
  } = props, restProps = __rest(props, [
    "accent",
    "align",
    "children",
    "muted",
    "size",
    "textOverflow",
    "weight"
  ]);
  let children = childrenProp;
  if (textOverflow === "ellipsis") {
    children = /* @__PURE__ */ React.createElement(SpanWithTextOverflow, null, children);
  }
  return /* @__PURE__ */ React.createElement(Root, __assign(__assign({
    "data-ui": "Text"
  }, restProps), {
    $accent: accent,
    $align: align,
    $muted: muted,
    ref,
    $size: size,
    $weight: weight
  }), children);
});
Text.displayName = "Text";
export {
  Text
};
