// sanity-plugin-dashboard-widget-netlify/dist/components/NetlifyWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _anchor = _interopRequireDefault(require("part:@sanity/components/buttons/anchor"));

var _NetlifyWidget = _interopRequireDefault(require("./NetlifyWidget.css"));

var _SiteList = _interopRequireDefault(require("./SiteList"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class NetlifyWidget extends _react.default.Component {
  render() {
    var netlifySitesUrl = 'https://app.netlify.com/account/sites';
    var _this$props = this.props,
        title = _this$props.title,
        description = _this$props.description,
        isLoading = _this$props.isLoading,
        sites = _this$props.sites,
        onDeploy = _this$props.onDeploy;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _NetlifyWidget.default.container
    }, /*#__PURE__*/_react.default.createElement("header", {
      className: _NetlifyWidget.default.header
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _NetlifyWidget.default.title
    }, title)), /*#__PURE__*/_react.default.createElement("div", {
      className: _NetlifyWidget.default.content
    }, description && /*#__PURE__*/_react.default.createElement("p", {
      className: _NetlifyWidget.default.description,
      dangerouslySetInnerHTML: {
        __html: description
      }
    }), /*#__PURE__*/_react.default.createElement(_SiteList.default, {
      isLoading: isLoading,
      onDeploy: onDeploy,
      sites: sites
    })), /*#__PURE__*/_react.default.createElement("div", {
      className: _NetlifyWidget.default.footer
    }, /*#__PURE__*/_react.default.createElement(_anchor.default, {
      disabled: isLoading,
      href: isLoading ? undefined : netlifySitesUrl,
      bleed: true,
      color: "primary",
      kind: "simple"
    }, "Manage sites at Netlify")));
  }

}

exports.default = NetlifyWidget;
