// @sanity/ui/dist/cjs/components/menu/menuButton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  MenuButton: () => MenuButton
});
var import_react = __toModule(require("react"));
var import_react_is = __toModule(require("react-is"));
var import_helpers = __toModule(require("../../helpers"));
var import_hooks = __toModule(require("../../hooks"));
var import_primitives = __toModule(require("../../primitives"));
const MenuButton = (0, import_react.forwardRef)(function MenuButton2(props, ref) {
  const {
    boundaryElement,
    button: buttonProp,
    id,
    menu: menuProp,
    placement,
    popoverScheme,
    portal,
    popover,
    popoverRadius,
    preventOverflow
  } = props;
  const [open, setOpen] = (0, import_react.useState)(false);
  const [focusLast, setFocusLast] = (0, import_react.useState)(false);
  const [buttonElement, setButtonElement] = (0, import_react.useState)(null);
  const [menuElement, setMenuElement] = (0, import_react.useState)(null);
  const handleButtonClick = (0, import_react.useCallback)(() => {
    setOpen((v) => !v);
    setFocusLast(false);
  }, []);
  const handleButtonKeyDown = (0, import_react.useCallback)((event) => {
    if (event.key === "ArrowDown") {
      event.preventDefault();
      setOpen(true);
      setFocusLast(false);
      return;
    }
    if (event.key === "ArrowUp") {
      event.preventDefault();
      setOpen(true);
      setFocusLast(true);
      return;
    }
  }, []);
  const handleMenuClickOutside = (0, import_react.useCallback)((event) => {
    const targetElement = (0, import_helpers.isHTMLElement)(event.target) ? event.target.closest("button") : null;
    if (targetElement !== buttonElement) {
      setOpen(false);
    }
  }, [buttonElement]);
  const handleMenuEscape = (0, import_react.useCallback)(() => {
    setOpen(false);
    if (buttonElement)
      buttonElement.focus();
  }, [buttonElement]);
  const handleBlur = (0, import_react.useCallback)((event) => {
    const target = event.relatedTarget;
    if (target === buttonElement) {
      return;
    }
    if ((0, import_helpers.isHTMLElement)(target) && !(menuElement == null ? void 0 : menuElement.contains(target))) {
      setOpen(false);
    }
  }, [buttonElement, menuElement]);
  const handleItemClick = (0, import_react.useCallback)(() => {
    setOpen(false);
    if (buttonElement)
      buttonElement.focus();
  }, [buttonElement]);
  (0, import_hooks.useClickOutside)((0, import_react.useCallback)(() => setOpen(false), []), [buttonElement, menuElement]);
  const menuProps = {
    "aria-labelledby": id,
    focusLast,
    onClickOutside: handleMenuClickOutside,
    onEscape: handleMenuEscape,
    onItemClick: handleItemClick,
    ref: setMenuElement,
    onBlurCapture: handleBlur
  };
  const menu = menuProp ? (0, import_react.cloneElement)(menuProp, menuProps) : null;
  const setButtonRef = (0, import_react.useCallback)((el) => {
    if (typeof ref === "function") {
      ref(el);
    } else if (ref) {
      ref.current = el;
    }
    setButtonElement(el);
  }, [ref]);
  const button = (0, import_react.useMemo)(() => import_react_is.default.isElement(buttonProp) ? (0, import_react.cloneElement)(buttonProp, {
    id,
    onClick: handleButtonClick,
    onKeyDown: handleButtonKeyDown,
    "aria-haspopup": true,
    "aria-expanded": open,
    ref: setButtonRef,
    selected: open
  }) : null, [buttonProp, handleButtonClick, handleButtonKeyDown, id, open, setButtonRef]);
  const popoverProps = (0, import_react.useMemo)(() => {
    return __assign({
      boundaryElement,
      placement,
      portal,
      radius: popoverRadius,
      overflow: preventOverflow,
      scheme: popoverScheme
    }, popover || {});
  }, [boundaryElement, placement, popover, popoverRadius, portal, preventOverflow, popoverScheme]);
  return /* @__PURE__ */ import_react.default.createElement(import_primitives.Popover, __assign(__assign({}, popoverProps), {
    content: menu,
    "data-ui": "MenuButton",
    open
  }), button || /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null));
});
