// @sanity/ui/dist/cjs/components/menu/menuItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  MenuItem: () => MenuItem
});
var import_react = __toModule(require("react"));
var import_react_is = __toModule(require("react-is"));
var import_hooks = __toModule(require("../../hooks"));
var import_primitives = __toModule(require("../../primitives"));
var import_useMenu = __toModule(require("./useMenu"));
const MenuItem = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    disabled,
    fontSize = 2,
    icon,
    iconRight,
    onClick,
    padding = 3,
    paddingX,
    paddingY,
    paddingTop,
    paddingRight,
    paddingBottom,
    paddingLeft,
    radius = 2,
    selected,
    space = 3,
    text
  } = props, restProps = __rest(props, [
    "children",
    "disabled",
    "fontSize",
    "icon",
    "iconRight",
    "onClick",
    "padding",
    "paddingX",
    "paddingY",
    "paddingTop",
    "paddingRight",
    "paddingBottom",
    "paddingLeft",
    "radius",
    "selected",
    "space",
    "text"
  ]);
  const {mount, onItemClick, onMouseEnter, onMouseLeave} = (0, import_useMenu.useMenu)();
  const rootRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => mount(rootRef.current, selected), [mount, selected]);
  const ref = (0, import_hooks.useForwardedRef)(forwardedRef);
  const handleClick = (0, import_react.useCallback)((event) => {
    if (disabled)
      return;
    if (onClick)
      onClick(event);
    if (onItemClick)
      onItemClick();
  }, [disabled, onClick, onItemClick]);
  const paddingProps = {
    padding,
    paddingX,
    paddingY,
    paddingTop,
    paddingRight,
    paddingBottom,
    paddingLeft
  };
  const setRef = (0, import_react.useCallback)((el) => {
    ref.current = el;
    rootRef.current = el;
  }, [ref]);
  return /* @__PURE__ */ import_react.default.createElement(import_primitives.Card, __assign(__assign({
    as: "button",
    "data-ui": "MenuItem"
  }, restProps), {
    disabled,
    onClick: handleClick,
    onMouseEnter,
    onMouseLeave,
    radius,
    ref: setRef,
    role: "menuitem",
    tabIndex: -1,
    type: "button"
  }), (icon || text || iconRight) && /* @__PURE__ */ import_react.default.createElement(import_primitives.Box, __assign({
    as: "span"
  }, paddingProps), /* @__PURE__ */ import_react.default.createElement(import_primitives.Flex, {
    as: "span"
  }, icon && /* @__PURE__ */ import_react.default.createElement(import_primitives.Text, {
    size: fontSize
  }, (0, import_react.isValidElement)(icon) && icon, (0, import_react_is.isValidElementType)(icon) && (0, import_react.createElement)(icon)), text && /* @__PURE__ */ import_react.default.createElement(import_primitives.Box, {
    flex: 1,
    marginLeft: icon ? space : void 0,
    marginRight: iconRight ? space : void 0
  }, /* @__PURE__ */ import_react.default.createElement(import_primitives.Text, {
    size: fontSize
  }, text)), iconRight && /* @__PURE__ */ import_react.default.createElement(import_primitives.Text, {
    size: fontSize
  }, (0, import_react.isValidElement)(iconRight) && iconRight, (0, import_react_is.isValidElementType)(iconRight) && (0, import_react.createElement)(iconRight)))), children);
});
MenuItem.displayName = "MenuItem";
