// @sanity/ui/dist/cjs/components/menu/menuGroup.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  MenuGroup: () => MenuGroup
});
var import_icons = __toModule(require("@sanity/icons"));
var import_react = __toModule(require("react"));
var import_styled_components = __toModule(require("styled-components"));
var import_primitives = __toModule(require("../../primitives"));
var import_menu = __toModule(require("./menu"));
var import_useMenu = __toModule(require("./useMenu"));
const Root = import_styled_components.default.button((props) => {
  const {theme} = props;
  const tone = theme.sanity.color.card;
  return import_styled_components.css`
    -webkit-font-smoothing: inherit;
    appearance: none;
    font: inherit;
    border: 0;
    border-radius: 0;
    background: none;
    color: inherit;
    text-align: left;
    margin: 0;
    padding: 0;
    outline: none;
    background-color: var(--card-bg-color);
    color: var(--card-fg-color);

    &:not(:disabled):focus {
      --card-bg-color: ${tone.selected.bg} !important;
      --card-fg-color: ${tone.selected.fg} !important;
    }

    &:not(:disabled):active {
      --card-bg-color: ${tone.pressed.bg};
      --card-fg-color: ${tone.pressed.fg};
    }

    &:disabled {
      --card-bg-color: ${tone.disabled.bg};
      --card-fg-color: ${tone.disabled.fg};
    }
  `;
});
const TextContainer = import_styled_components.default.span`
  &:not([hidden]) {
    display: flex;
  }

  & > div:first-child {
    flex: 1;
    min-width: 0;
  }
`;
function MenuGroup({children, title}) {
  const [open, setOpen] = (0, import_react.useState)(false);
  const {mount, onItemClick, onMouseEnter, onMouseLeave} = (0, import_useMenu.useMenu)();
  const rootRef = (0, import_react.useRef)(null);
  const mouseLeaveTimeoutRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => mount(rootRef.current), [mount]);
  const handleMouseEnter = (event) => {
    if (mouseLeaveTimeoutRef.current) {
      clearTimeout(mouseLeaveTimeoutRef.current);
    }
    onMouseEnter(event);
    setOpen(true);
  };
  const handleMouseLeave = (event) => {
    onMouseLeave(event);
    mouseLeaveTimeoutRef.current = setTimeout(() => {
      setOpen(false);
    }, 5e3);
  };
  const handleMenuMouseEnter = () => {
    if (mouseLeaveTimeoutRef.current !== null) {
      clearTimeout(mouseLeaveTimeoutRef.current);
      mouseLeaveTimeoutRef.current = null;
    }
  };
  const handleMenuMouseLeave = () => {
    mouseLeaveTimeoutRef.current = setTimeout(() => {
      setOpen(false);
    }, 5e3);
  };
  (0, import_react.useEffect)(() => {
    return () => {
      if (mouseLeaveTimeoutRef.current !== null) {
        clearTimeout(mouseLeaveTimeoutRef.current);
        mouseLeaveTimeoutRef.current = null;
      }
    };
  }, []);
  const content = /* @__PURE__ */ import_react.default.createElement(import_menu.Menu, {
    onMouseEnter: handleMenuMouseEnter,
    onMouseLeave: handleMenuMouseLeave
  }, children);
  return /* @__PURE__ */ import_react.default.createElement(import_primitives.Popover, {
    content,
    "data-ui": "MenuGroup",
    open,
    placement: "right"
  }, /* @__PURE__ */ import_react.default.createElement(Root, {
    "data-ui": "MenuGroup",
    onClick: onItemClick,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    ref: rootRef
  }, /* @__PURE__ */ import_react.default.createElement(import_primitives.Box, {
    padding: 3
  }, /* @__PURE__ */ import_react.default.createElement(TextContainer, null, /* @__PURE__ */ import_react.default.createElement(import_primitives.Text, null, title), /* @__PURE__ */ import_react.default.createElement(import_primitives.Text, null, /* @__PURE__ */ import_react.default.createElement(import_icons.ChevronRightIcon, null))))));
}
