// @sanity/ui/dist/cjs/components/menu/menu.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __assign = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  Menu: () => Menu
});
var import_react = __toModule(require("react"));
var import_styled_components = __toModule(require("styled-components"));
var import_hooks = __toModule(require("../../hooks"));
var import_primitives = __toModule(require("../../primitives"));
var import_utils = __toModule(require("../../utils"));
var import_helpers = __toModule(require("./helpers"));
var import_menuContext = __toModule(require("./menuContext"));
const Root = (0, import_styled_components.default)(import_primitives.Box)`
  outline: none;
`;
const Menu = (0, import_react.forwardRef)((props, ref) => {
  const {
    children,
    focusLast,
    onClickOutside,
    onEscape,
    onItemClick,
    onItemSelect,
    padding = 1,
    space = 1
  } = props, restProps = __rest(props, [
    "children",
    "focusLast",
    "onClickOutside",
    "onEscape",
    "onItemClick",
    "onItemSelect",
    "padding",
    "space"
  ]);
  const {isTopLayer} = (0, import_utils.useLayer)();
  const [rootElement, setRootElement] = (0, import_react.useState)(null);
  const itemsRef = (0, import_react.useRef)([]);
  const [activeIndex, setActiveIndex] = (0, import_react.useState)(-1);
  const [activeElement, setActiveElement] = (0, import_react.useState)(null);
  const activeElementRef = (0, import_react.useRef)(activeElement);
  (0, import_react.useEffect)(() => {
    if (onItemSelect)
      onItemSelect(activeIndex);
  }, [activeIndex, onItemSelect]);
  (0, import_react.useEffect)(() => {
    activeElementRef.current = activeElement;
  }, [activeElement]);
  (0, import_react.useEffect)(() => {
    const rafId = window.requestAnimationFrame(() => {
      if (rootElement) {
        if (activeElementRef.current) {
          activeElementRef.current.focus();
          setActiveIndex(itemsRef.current.indexOf(activeElementRef.current));
          return;
        }
        const element = itemsRef.current[focusLast ? itemsRef.current.length - 1 : 0];
        if (element) {
          element.focus();
          setActiveIndex(itemsRef.current.indexOf(element));
        }
      }
    });
    return () => {
      window.cancelAnimationFrame(rafId);
    };
  }, [focusLast, rootElement]);
  const setRef = (0, import_react.useCallback)((el) => {
    setRootElement(el);
    if (typeof ref === "function")
      ref(el);
    else if (ref)
      ref.current = el;
  }, [ref]);
  const mount = (0, import_react.useCallback)((element, selected) => {
    if (!element)
      return () => void 0;
    if (!itemsRef.current.includes(element)) {
      itemsRef.current.push(element);
    }
    if (selected === true) {
      setActiveElement(element);
    }
    return () => {
      const idx = itemsRef.current.push(element);
      if (idx > -1) {
        itemsRef.current.splice(idx, 1);
      }
    };
  }, []);
  const handleKeyDown = (0, import_react.useCallback)((event) => {
    if (event.key === "ArrowUp") {
      event.preventDefault();
      const focusableElements = (0, import_helpers.getFocusableElements)(itemsRef.current);
      const focusableLen = focusableElements.filter(({focusable: focusable2}) => focusable2).length;
      if (focusableLen === 0)
        return;
      const len = focusableElements.length;
      let currentIndex = activeIndex;
      let focusable = false;
      let element = null;
      while (!focusable) {
        currentIndex = (currentIndex - 1 + len) % len;
        element = focusableElements[currentIndex].element;
        focusable = focusableElements[currentIndex].focusable;
      }
      setActiveIndex(currentIndex);
      if (element)
        element.focus();
      return;
    }
    if (event.key === "ArrowDown") {
      event.preventDefault();
      const focusableElements = (0, import_helpers.getFocusableElements)(itemsRef.current);
      const focusableLen = focusableElements.filter(({focusable: focusable2}) => focusable2).length;
      if (focusableLen === 0)
        return;
      const len = focusableElements.length;
      let currentIndex = activeIndex;
      let focusable = false;
      let element = null;
      while (!focusable) {
        currentIndex = (currentIndex + 1) % len;
        element = focusableElements[currentIndex].element;
        focusable = focusableElements[currentIndex].focusable;
      }
      setActiveIndex(currentIndex);
      if (element)
        element.focus();
      return;
    }
  }, [activeIndex]);
  const handleItemMouseEnter = (0, import_react.useCallback)((event) => {
    const element = event.currentTarget;
    setActiveIndex(itemsRef.current.indexOf(element));
    element.focus();
  }, []);
  const handleItemMouseLeave = (0, import_react.useCallback)(() => {
    setActiveIndex(-1);
    rootElement == null ? void 0 : rootElement.focus();
  }, [rootElement]);
  (0, import_hooks.useClickOutside)((0, import_react.useCallback)((event) => isTopLayer && onClickOutside && onClickOutside(event), [
    isTopLayer,
    onClickOutside
  ]), [rootElement]);
  (0, import_hooks.useGlobalKeyDown)((0, import_react.useCallback)((event) => {
    if (!isTopLayer)
      return;
    if (event.key === "Escape") {
      event.stopPropagation();
      if (onEscape)
        onEscape();
    }
  }, [isTopLayer, onEscape]));
  const value = (0, import_react.useMemo)(() => ({
    version: 0,
    activeElement,
    activeIndex,
    mount,
    onMouseEnter: handleItemMouseEnter,
    onMouseLeave: handleItemMouseLeave,
    onItemClick
  }), [activeElement, activeIndex, mount, handleItemMouseEnter, handleItemMouseLeave, onItemClick]);
  return /* @__PURE__ */ import_react.default.createElement(import_menuContext.MenuContext.Provider, {
    value
  }, /* @__PURE__ */ import_react.default.createElement(Root, __assign(__assign({
    "data-ui": "Menu"
  }, restProps), {
    onKeyDown: handleKeyDown,
    padding,
    ref: setRef,
    role: "menu",
    tabIndex: -1
  }), /* @__PURE__ */ import_react.default.createElement(import_primitives.Stack, {
    space
  }, children)));
});
Menu.displayName = "Menu";
